source/merge.ts

```diff
@@ -0,0 +1,93 @@
+import {URL} from 'url';
+import is from '@sindresorhus/is';
+// XO doesn't recognize `export type` for some reason.
+// eslint-disable-next-line import/named
+import knownHookEvents, {Hooks, HookType, HookEvent} from './known-hook-events';
+
+// TODO: Use the Got options-object types.
+interface Options {
+	hooks?: Partial<Hooks>;
+	[key: string]: unknown | Options;
+}
+
+export default function merge(target: Options, ...sources: Options[]) {
+	for (const source of sources) {
+		for (const [key, sourceValue] of Object.entries(source)) {
+			if (is.undefined(sourceValue)) {
+				continue;
+			}
+
+			const targetValue = target[key];
+			if (is.urlInstance(targetValue) && (is.urlInstance(sourceValue) || is.string(sourceValue))) {
+				target[key] = new URL(sourceValue as string, targetValue);
+			} else if (is.plainObject(sourceValue)) {
+				if (is.plainObject(targetValue)) {
+					target[key] = merge({}, targetValue as Options, sourceValue as Options);
+				} else {
+					target[key] = merge({}, sourceValue as Options);
+				}
+			} else if (is.array(sourceValue)) {
+				target[key] = merge([], sourceValue);
+			} else {
+				target[key] = sourceValue;
+			}
+		}
+	}
+
+	return target;
+}
+
+export function mergeOptions(...sources: Options[]) {
+	sources = sources.map(source => source || {});
+	const merged = merge({}, ...sources);
+
+	// TODO: This is a funky situation. Even though we "know" that we're going to
+	//       populate the `hooks` object in the loop below, TypeScript want us to //       put them into the object upon initialization, because it cannot infer
+	//       that they are going to conform correctly in runtime.
+	// eslint-disable-next-line @typescript-eslint/no-object-literal-type-assertion
+	const hooks = {} as { [Key in HookEvent]: HookType[] };
```

Comment:
This one is quirky and probably not correct, but I have tried multiple ways to reflect that we _know_ that it's going to be populated later, in the loop underneath but we don't want to give it the values upfront.

Reply:
I'm not a TypeScript master either :P

Is there maybe another way? For example the [Hooks interface](https://github.com/sindresorhus/got/pull/714/files#diff-8cbb1059c76170f4b0419c6df7f197baR18):

```ts
const hooks: Hooks = {}
```

Reply:
ping?

Reply:
I have tried that and also with `as Hooks`.

```ts
hooks[hook] = hooks[hook].concat(source.hooks[hook] || []);
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                         [ts] Cannot invoke an expression whose type lacks a call signature.
                         Type '{ (...items: ConcatArray<InitHook>[]): InitHook[]; (...items: (InitHook | ConcatArray<InitHook>)[]): InitHook[]; } | { (...items: ConcatArray<BeforeRequestHook>[]): BeforeRequestHook[]; (...items: (BeforeRequestHook | ConcatArray<...>)[]): BeforeRequestHook[]; } | { ...; } | { ...; } | { ...; } | { ...; }'
                         has no compatible call signatures. [2349]
```

Seen as this line is temporary, I don't think its much of an issue just an annoyance.

Reply:
Oh, thanks for the explaination, it's clear now. I'll try to think of something...